### PR TITLE
Result formatting

### PR DESCRIPTION
**Pull request by AronBuzogany**
This PR implements the following:

- Possibility to specify the type of a variable   
- Possibility to define a representation for the results of a querry corresponding to a variable with a certain type.
- Possibility to define a comperator which allows a user to sort a column corresponding to a variable with a certain type.
- Extends the README file with instructions of how to add or edit the representations and comperators.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,4 +37,44 @@ The configuration file follows a simple structure.
...
+
+### Adding variable type
+
+You can specify the type of a variable by extending its name with the type in the query as such: ```variableName_variableType```. The underscore ```_``` here is crutial to make a clear distinction between name and type. 
+
+### Representation Mapper 
+
+If you want to add your own type representations you can do this by adding your representation to the [typeMapper.js](./src/typeMapper.js) file. 
+
+The mapper follows a simple structure:
```
Comment (pheyvaer) [inline README.md:50]: No need for the word "simple" here.

File: README.md [hunk 2]
```diff
@@ -22,7 +22,7 @@ Which will start the web application
 
 The configuration file follows a simple structure. 
 
-```json
+```js
```
Comment (pheyvaer) [inline README.md:25]: Why is this changed?

File: README.md [hunk 3]
```diff
@@ -37,4 +37,44 @@ The configuration file follows a simple structure.
         ...
     ]
 }
-```
\ No newline at end of file
+```
+
+### Adding variable type
+
+You can specify the type of a variable by extending its name with the type in the query as such: ```variableName_variableType```. The underscore ```_``` here is crutial to make a clear distinction between name and type.
```
Comment (pheyvaer) [inline README.md:44]: - Explain why someone would want to add the type.
- Give an example.

File: README.md [hunk 4]
```diff
@@ -37,4 +37,44 @@ The configuration file follows a simple structure.
...
+```js
+{
+    "typeName": mapperFunction,
+    ... 
+}
+```
+
+With ```typeName``` being the name of the variable as defined in the query and ```mapperFunction``` a function that takes the query result for the corresponding variable and returns either a string or a [React](https://react.dev/) component (see below).  
+
+Examples of how you can do this can already be found in that same file.
```
Comment (pheyvaer) [inline README.md:61]: - Paragraph of one sentence are, in general, bad practice.
- Provide a link to where exactly in the file that is.

File: README.md [hunk 5]
```diff
@@ -37,4 +37,44 @@ The configuration file follows a simple structure.
...
+    "typeName": mapperFunction,
+    ... 
+}
+```
+
+With ```typeName``` being the name of the variable as defined in the query and ```mapperFunction``` a function that takes the query result for the corresponding variable and returns either a string or a [React](https://react.dev/) component (see below).  
+
+Examples of how you can do this can already be found in that same file. 
+
+The web application uses [gridjs-react](https://gridjs.io/docs/integrations/react) internally. This allows us to add [html](https://nl.wikipedia.org/wiki/HyperText_Markup_Language) or [React](https://react.dev/) components as representations of the variable. To do this you need to import ```_``` function from [gridjs-react module](https://www.npmjs.com/package/gridjs-react) and call it with the component as its variable. An example for this is also already provided.
```
Comment (pheyvaer) [inline README.md:63]: - Newlines should be added after a sentence is finished.
- Add a link to the example.

File: README.md [hunk 6]
```diff
@@ -37,4 +37,44 @@ The configuration file follows a simple structure.
...
\ No newline at end of file
+```
+
+### Adding variable type
+
+You can specify the type of a variable by extending its name with the type in the query as such: ```variableName_variableType```. The underscore ```_``` here is crutial to make a clear distinction between name and type. 
+
+### Representation Mapper 
+
+If you want to add your own type representations you can do this by adding your representation to the [typeMapper.js](./src/typeMapper.js) file.
```
Comment (pheyvaer) [inline README.md:48]: Why would I want to add my own type representations? Can you give an example?

File: src/typeMapper.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    return value 
+  }
+}
+
+/**
+ * 
+ * @param {String} value 
+ * @returns Gets the value out of a literal as defined in https://www.w3.org/TR/rdf11-concepts/ , thus the value without type, language specifics...
+ */
+function getLiteralValue(value){
```
Comment (pheyvaer) [inline src/typeMapper.js:54]: Comunica returns [RDF/JS](https://rdf.js.org/data-model-spec/) model objects, so getting the value should be done through `Literal.value`.

File: src/typeMapper.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * 
+ * @param {Number} value string value which is to be converted to a number
+ * @param {Function} parseFunction a function that takes the string value of a number and converts it to the corresponding number subclass 
+ * @returns A number parsed in the correct type (e.g. integer, float...) or an empty string (instead of NaN)
+ */
+function literalToNumber(value, parseFunction){
+  let literalValue = getLiteralValue(value)
+  if(literalValue){
```
Comment (pheyvaer) [inline src/typeMapper.js:41]: What happens if the literal value is `""`?

**Review by pheyvaer — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (AronBuzogany) [reply README.md:25]: The 3 dots symbolizing that the object/list contains more elements is not json syntax, so it gets marked red in markdown. It is however javascript syntax.

File: src/config.json [hunk 9]
```diff
@@ -15,6 +15,13 @@
             "description": "A second simple test query",
             "id": "456",
             "sources": ["https://ruben.verborgh.org/profile/#me"]
+        },
+        {
+            "queryLocation": "images_test.ttl",
+            "name": "A Test For Images",
+            "description": "A small test to test the functionality of the web-app concerning images",
```
[the remark this answers is not in the document]
Comment (AronBuzogany) [reply src/config.json]: This is resolved in [commit 48a8097](https://github.com/SolidLabResearch/generic-data-viewer/pull/6/commits/48a80972bebdd3ae267b795e3282ec8557128130)

File: README.md (hunk 1 above)
Comment (AronBuzogany) [reply README.md]: This is resolved in [commit 259d097](https://github.com/SolidLabResearch/generic-data-viewer/pull/6/commits/259d0979b5cee6cb7d7348032e3ab3404323d93a)

File: README.md (hunk 3 above)
Comment (AronBuzogany) [reply README.md:48]: This is resolved in [commit 259d097](https://github.com/SolidLabResearch/generic-data-viewer/pull/6/commits/259d0979b5cee6cb7d7348032e3ab3404323d93a)

File: README.md (hunk 4 above)
Comment (AronBuzogany) [reply README.md]: This is resolved in [commit 259d097](https://github.com/SolidLabResearch/generic-data-viewer/pull/6/commits/259d0979b5cee6cb7d7348032e3ab3404323d93a)

File: README.md (hunk 6 above)
Comment (AronBuzogany) [reply README.md:52]: I have added both in [commit 259d097](https://github.com/SolidLabResearch/generic-data-viewer/pull/6/commits/259d0979b5cee6cb7d7348032e3ab3404323d93a)

File: src/typeMapper.js (hunk 7 above)
Comment (AronBuzogany) [reply src/typeMapper.js]: This is indeed much easier and seems way more efficient, I have adjusted it respectively in [commit bf63403](https://github.com/SolidLabResearch/generic-data-viewer/pull/6/commits/bf634037b62fff67c6ed4092087ad7a5b38063ce)

File: src/typeMapper.js (hunk 8 above)
Comment (pheyvaer) [reply src/typeMapper.js:41]: But the function claims to return a number and string is not a number. So either the function doc needs to change or the function should throw an error.

Comment (AronBuzogany) [reply src/typeMapper.js:41]: I have changed this to return undefined instead of an empty string in [commit 85429a3](https://github.com/SolidLabResearch/generic-data-viewer/pull/6/commits/85429a36a3fe7d215d11bfad0ac3094c68afed17)

File: README.md (hunk 3 above)
Comment (pheyvaer) [reply README.md]: > either a URL or a literal value

This statement is incorrect. You can also get a blank node https://www.w3.org/TR/rdf12-concepts/#section-blank-nodes

Comment (AronBuzogany) [reply README.md]: Thank you, I have adjusted this in [commit 16ad860](https://github.com/SolidLabResearch/generic-data-viewer/pull/6/commits/16ad8604d233a32e95dd04e969fb48197f2bb0d1)

**Review by pheyvaer — CHANGES_REQUESTED** [review]
The grid flashes when new entries are added. It should be possible to add new data to the grid without having this.

**Comment by AronBuzogany** [comment]
> The grid flashes when new entries are added. It should be possible to add new data to the grid without having this.

This was indeed a css issue on my end, I have now solved this in [commit b4f898c](https://github.com/SolidLabResearch/generic-data-viewer/pull/6/commits/b4f898cffbd4e8fae9d0815a36488be4a43dc703).
